maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/AbstractSurefireMojo.java

```diff
@@ -319,25 +320,49 @@ public abstract class AbstractSurefireMojo extends AbstractMojo implements Suref
      */
     @Deprecated
     @Parameter
-    private Properties systemProperties;
+    Properties systemProperties;
 
     /**
      * List of System properties to pass to a provider.
      * The effective system properties given to a provider are contributed from several sources:
      * <ol>
+     * <li>properties set via {@link #argLine} with {@code -D} (only for forked executions)</li>
      * <li>{@link #systemProperties}</li>
      * <li>{@link AbstractSurefireMojo#getSystemPropertiesFile()} (set via parameter {@code systemPropertiesFile} on some goals)</li>
      * <li>{@link #systemPropertyVariables}</li>
-     * <li>User properties from {@link MavenSession#getUserProperties()}, usually set via CLI options given with {@code -D}</li>
+     * <li>User properties from {@link MavenSession#getUserProperties()}, usually set via CLI options given with {@code -D} on the current Maven process</li>
+     * <li>{@link #userPropertyVariables}</li>
      * </ol>
      * Later sources may overwrite same named properties from earlier sources, that means for example that one cannot overwrite user properties with either
-     * {@link #systemProperties}, {@link AbstractSurefireMojo#getSystemPropertiesFile()} or {@link #systemPropertyVariables}.
+     * {@link #systemProperties}, {@link #getSystemPropertiesFile()} or {@link #systemPropertyVariables} but only with
+     * {@link #userPropertyVariables}.
+     * <p>
+     * Certain properties may only be overwritten via {@link #argLine} (applicable only for forked executions) because their values are cached and only evaluated at the start of the JRE.
+     * Those include:
+     * <ul>
+     * <li>{@code java.library.path}</li>
+     * <li>{@code file.encoding}</li>
+     * <li>{@code jdk.map.althashing.threshold}</li>
+     * <li>{@code line.separator}</li>
+     * </ul>
      *
      * @since 2.5
      * @see #systemProperties
      */
     @Parameter
-    private Map<String, String> systemPropertyVariables;
+    Map<String, String> systemPropertyVariables;
+
+    /**
+     * List of user properties to pass to a provider.
+     * Similar to {@link #systemPropertyVariables} but having a higher precedence, therefore allows to overwrite user properties from the current Maven session.
+     * This should only be used in case a user property from the parent process needs to be explicitly overwritten.
+     * Regular properties should be set via {@link #systemPropertyVariables} instead in order to allow them to be overwritten
+     * via CLI arguments ({@code -Dmyproperty=myvalue})
+     * @since 3.4
+     * @see #systemPropertyVariables
+     */
+    @Parameter
+    Map<String, String> userPropertyVariables;
```

Comment:
From test perspective there are system properties - not a user properties.

Next parameter like this can confusing user and will be used in random.

I would like to consider changing a priority for `systemPropertyVariables`, 
even more user can take old behavior by settings properties in project and use it in configuration.

Maybe we should don't pass a `MavenSession#getUserProperties()` by default to tests ....

Reply:
> Maybe we should don't pass a MavenSession#getUserProperties() by default to tests ....

That calls for backwards compatibility issues for certain edge cases. I don't think we can do that for a 3.x release.

Reply:
Any suggestion for another name? As I said above, I don't think we can fix this in a backwards-compatible way without introducing another parameter...

Reply:
@kwin The let's make it right -- after all -- with a deprecated param with retains old behavior. Surefire 4 is far away, this issue has been open for too long.

Reply:
> let's make it right

What is right? As outlined in https://issues.apache.org/jira/browse/SUREFIRE-1385?focusedCommentId=17864170&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17864170 there is no simple solution IMHO: 
The default should still be to allow overwrites with CLI arguments. Overwriting CLI argument is the edge case which we need to support on an exceptional base. I don't see a simple solution without adding a param. Please make a proposal what you consider right here.

Reply:
My abstract understanding:
* system properties (include MAVEN_OPTS logically)
* system properties file
* system property variables in POM
* user properties file
* user property variables in POM
* user porperties from CLI

WDYT?

Reply:
Do you describe the to-be state? This is a lot different from the as is and not really clear which of those sources come from mojo parameters (with which name). Also there are even more sources considered than today….

Reply:
My abstract expectation regardless of the current state.

Reply:
Not sure how this helps here. Let us focus on how you think it should work!

Reply:
What about a boolean parameter `includeUserPropertiesInSystemProperties` which is `true` by default (for backwards compatibility). If set to `false` this will prevent passing on the user properties to the provider at all. This will make the addition `userPropertyVariables` parameter obsolete from my perspective.

Reply:
I implemented the aforementioned approach now in https://github.com/apache/maven-surefire/pull/762/commits/f8b0f57720feafe8d84103f1327c0a8482ed3ce3.

Reply:
OK, I will try to dedicate some to it soon.

Reply:
@slawekjaranowski I agree with you that passing user properties by default is a bad idea.